### PR TITLE
adding types to theme

### PR DESCRIPTION
**Pull request by spenc53**
Signed-off-by: Spencer Seeger <seeger.spencer@gmail.com>

#### What does this PR do?
Adds types to the themes

#### Where should the reviewer start?
Start by checking to make sure that the theme definition is correct. I copied from index.d.ts but, had to combine for color.

#### What testing has been done on this PR?
I ran through the tutorial and made sure everything was working expected. I ran the test cases. I also opened up Visual Studio Code and made sure that the types were appearing correctly.

#### How should this be manually tested?
Open up in a code editor and make sure that the types appear.

#### Any background context you want to provide?
N/A

#### What are the relevant issues?
https://github.com/grommet/grommet/issues/3278

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
Yes, it will let users know that themes are now types

#### Is this change backwards compatible or is it a breaking change?
It is backwards compatible

### CODE REVIEW

File: src/js/themes/base.d.ts [hunk 1]
```diff
@@ -0,0 +1,825 @@
+import {DeepReadonly} from '../utils'
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]:
```suggestion
import { DeepReadonly } from '../utils'
```

File: src/js/themes/base.d.ts [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+export declare const generate: (baseSpacing = 24, scale = 6) => DeepReadonly<ThemeType>;
+
+declare const colors: {
+    active: string;
+    black: string;
+    border: {
+        dark: string;
+        light: string;
+    };
+    brand: string;
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]: I think all colors should be able to accept the ColorType from https://github.com/grommet/grommet/blob/8b25e3abd6fc42ec9977053c750ccb7854395d55/src/js/utils/index.d.ts#L43

In the sandbox example I created, I'm showing how brand color isn't necessarily a string, but can also accept an object.
https://codesandbox.io/s/colors-i19hp

File: src/js/themes/index.d.ts [hunk 3]
```diff
@@ -1,1604 +1,3 @@
...
-      hover: {
-          color: string;
-      };
-      place: {
-          active: string;
-          base: string;
-      };
-  };
-}>;
+export { ThemeType, base, generate } from './base';
```
Comment (ShimiSun) [inline src/js/themes/index.d.ts]: This is a great cleanup, thanks!

File: src/js/themes/base.d.ts [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+        responsiveBreakpoint: string;
+    };
+    button: {
+        border: {
+            width: string;
+            radius: string;
+        };
+        primary: {};
+        disabled: {
+            opacity: number;
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]:
```suggestion
            opacity: number | string;
```
e.g. `opacity="strong"`
I think it is safe to mention it for all opacity styles in this file.

File: src/js/themes/base.d.ts [hunk 5]
```diff
@@ -0,0 +1,825 @@
...
+            };
+            width: string;
+        };
+        check: {
+            extend: string;
+            radius: string;
+            thickness: string;
+        };
+        extend: string;
+        color: string | { dark: string; light: string; };
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]: can we refactor this to use ColorType https://github.com/grommet/grommet/blob/master/src/js/utils/index.d.ts#L43

File: src/js/themes/base.d.ts [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+        icon: { size: string; extend: string };
+        icons: {};
+        hover: {
+            border: {
+                color: {
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]: this should be able to get a string as well as an object.

File: src/js/themes/base.d.ts [hunk 7]
```diff
@@ -0,0 +1,825 @@
...
+              dark: string;
+              light: string;
+            };
+            radius: string;
+            size: string;
+            knob: { extend: string };
+        };
+    };
+    clock: {
+        analog: {
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]: we try to use 2 space indentations.

File: src/js/themes/base.d.ts (hunk 4 above)
Comment (spenc53) [reply src/js/themes/base.d.ts]: Is there a type defined for opacity somewhere? It might be useful to have if not.

Comment (ShimiSun) [reply src/js/themes/base.d.ts]: you can add one to grommet/src/js/utils/index.d.ts if it's not there.

Comment (spenc53) [reply src/js/themes/base.d.ts]: Awesome! Should I use the values from this page (https://v2.grommet.io/rangeselector) and number as well?

Comment (ShimiSun) [reply src/js/themes/base.d.ts]: yep 💯

File: src/js/utils/index.d.ts [hunk 8]
```diff
@@ -47,3 +47,4 @@ export type KeyboardType = ((event: React.KeyboardEvent<HTMLElement>) => void);
 export type MarginType = "none" | "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | {bottom?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,horizontal?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,left?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,right?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,top?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string,vertical?: "xxsmall" | "xsmall" | "small" | "medium" | "large" | "xlarge" | string} | string;
 export type PlaceHolderType = string | JSX.Element | React.ReactNode;
 export type TextAlignType = "start" | "center" | "end";
+export type OpacityType = "weak" | "medium" | "strong" | string | true | false | number;
```
Comment (ShimiSun) [inline src/js/utils/index.d.ts]: Let's move this type to line #48, so we will be consistent with the existing alphabetical order.

File: src/js/themes/base.d.ts [hunk 9]
```diff
@@ -0,0 +1,762 @@
...
+  control: ColorType;
+  focus: ColorType;
+  placeholder: ColorType;
+  selected: ColorType;
+  text: ColorType;
+  icon: ColorType;
+  white: ColorType;
+};
+  
+type Colors = typeof colors & {
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]: all these colors should also support ColorType structure, https://codesandbox.io/s/accent-1-i19hp

File: src/js/themes/base.d.ts [hunk 10]
```diff
@@ -0,0 +1,762 @@
...
+      padding: string;
+      weight: number;
+    };
+    opacity: {
+      strong: number;
+      medium: number;
+      weak: number;
+    };
+    selected: {
+      background: string;
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]: I think there is a BackgroundType that could be used here.

File: src/js/themes/base.d.ts [hunk 11]
```diff
@@ -0,0 +1,762 @@
...
+  };
+  box: {
+    responsiveBreakpoint: string;
+  };
+  button: {
+    border: {
+      width: string;
+      radius: string;
+    };
+    primary: {};
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]:
```suggestion
    primary: ColorType;
```

File: src/js/themes/base.d.ts [hunk 12]
```diff
@@ -0,0 +1,762 @@
...
+    disabled: {
+      opacity: number;
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts]:
```suggestion
      opacity: OpacityType;
```

**Pull request merged**